pkg/api/errors/pubsub.go

```diff
@@ -17,170 +17,229 @@ import (
-	PostFixPublishMessage          = "PUBLISH_MESSAGE"
-	PostFixCloudEventCreation      = "CLOUD_EVENT_CREATION"
-	PostFixMarshalEnvelope         = "MARSHAL_ENVELOPE"
-	PostFixMarshalEvents           = "MARSHAL_EVENTS"
-	PostFixUnMarshalEvents         = "UNMARSHAL_EVENTS"
-	PostFixForbidden               = "FORBIDDEN"
-	PostFixOutbox                  = "OUTBOX"
-)
+type PubSubError struct {
+	name string
+}
 
-func PubSubNameEmpty(name string, pubsubType string, metadata map[string]string) error {
-	message := "pubsub name is empty"
-	return kiterrors.NewBuilder(
-		grpcCodes.InvalidArgument,
-		http.StatusNotFound,
-		message,
-		"ERR_PUBSUB_EMPTY",
-	).
-		WithErrorInfo(kiterrors.CodePrefixPubSub+PostFixNameEmpty, metadata).
-		WithResourceInfo(pubsubType, name, "", message).
-		Build()
+type PubSubMetadataError struct {
+	p                *PubSubError
+	metadata         map[string]string
+	skipResourceInfo bool
 }
 
-func PubSubTopicEmpty(name string, pubsubType string, metadata map[string]string) error {
-	message := fmt.Sprintf("topic is empty in pubsub %s", name)
-	return kiterrors.NewBuilder(
-		grpcCodes.InvalidArgument,
-		http.StatusNotFound,
-		message,
-		"ERR_TOPIC_EMPTY",
-	).
-		WithErrorInfo(kiterrors.CodePrefixPubSub+InFixTopic+PostFixNameEmpty, metadata).
-		WithResourceInfo(pubsubType, name, "", message).
-		Build()
+type PubSubTopicError struct {
+	p     *PubSubMetadataError
+	topic string
 }
 
-func PubSubMetadataDeserialize(name string, pubsubType string, metadata map[string]string, err error) error {
-	message := fmt.Sprintf("failed deserializing metadata. Error: %s", err.Error())
-	return kiterrors.NewBuilder(
-		grpcCodes.InvalidArgument,
-		http.StatusBadRequest,
-		message,
-		"ERR_PUBSUB_REQUEST_METADATA",
-	).
-		WithErrorInfo(kiterrors.CodePrefixPubSub+PostFixMetadataDeserialization, metadata).
-		WithResourceInfo(pubsubType, name, "", message).
-		Build()
+func PubSub(name string) *PubSubError {
+	return &PubSubError{
+		name: name,
+	}
+}
+
+func (p *PubSubError) WithMetadata(metadata map[string]string) *PubSubMetadataError {
+	return &PubSubMetadataError{
+		p:        p,
+		metadata: metadata,
+	}
+}
+
+func (p *PubSubError) WithAppError(appID string, err error) *PubSubMetadataError {
+	meta := map[string]string{
+		"appID": appID,
+	}
+	if err != nil {
+		meta["error"] = err.Error()
+	}
+	return &PubSubMetadataError{
+		p:        p,
+		metadata: meta,
+	}
+}
+
+func (p *PubSubError) withTopicError(topic string, err error) *PubSubMetadataError {
+	return &PubSubMetadataError{
+		p: p,
+		metadata: map[string]string{
+			"topic": topic,
+			"error": err.Error(),
+		},
+	}
 }
 
-func PubSubPublishMessage(name string, pubsubType string, topic string, err error) error {
-	message := fmt.Sprintf("error when publishing to topic %s in pubsub %s: %s", topic, name, err.Error())
-	return kiterrors.NewBuilder(
-		grpcCodes.Internal,
+func (p PubSubError) PublishMessage(topic string, err error) error {
```

Comment:
I really like the "fluent-like" way this refactor builds errors. I'm a bit divided about the `PubSubMetadataError` in the middle, but I understand why it's there. It has its pros and cons.
It just feels a bit weird to me to have the intermediate methods that return `PubSubMetadataError` for some errors, but not for others. Right now we have final methods (the one that calls `build`) on both `PubSubError` and `PubSubMetadataError` and that might be a bit confusing when you see it as an IDE suggestion.

Reply:
How should be proceed here, remove the the `Metadata` intermediate struct?

Reply:
Since the metadata struct is very useful for setting the appid metadata, I think we can just leave it for now. If we see the relationships between the three (PubSubError, PubSubMetadataError and PubSubTopicError) get more complex in the future, we can always refactor.